server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/core/executor/execution/model/ExecutionGroupContext.java

```diff
@@ -0,0 +1,138 @@
+        this.id2ExecutionUnit = executionGroups.stream().flatMap(group -> group.getExecutionUnits().stream())
+                .collect(Collectors.toMap(ExecutionUnit::getId, unit -> unit));
+    }
+
+    public void execute() throws InterruptedException {
+        if (CollectionUtils.isEmpty(executionGroups)) {
+            return;
+        }
+        executionGroups.forEach(group -> group.getExecutionUnits().forEach(unit -> unit.beforeExecute(this)));
+        for (ExecutionGroup<T, R> group : executionGroups) {
+            if (Thread.currentThread().isInterrupted()) {
+                Thread.currentThread().interrupt();
+                throw new InterruptedException();
+            }
+            group.execute(this.executorService, this);
+            waitForCompletion(group.getExecutionUnits().stream().map(ExecutionUnit::getId).collect(Collectors.toSet()));
+            completedGroupCount++;
+        }
+    }
+
+    public ExecutionResult<R> getExecutionResult(String executionId) {
+        return executionId2Result.get(executionId);
+    }
+
+    public void setExecutionResult(String executionId,
+            BiFunction<String, ExecutionResult<R>, ExecutionResult<R>> result) {
+        executionId2Result.compute(executionId, result);
+    }
+
+    public boolean isCompleted() {
+        return executionId2Result.values().stream().allMatch(ExecutionResult::isCompleted);
+    }
+
+    public Map<String, ExecutionResult<R>> getResults() {
+        return executionId2Result;
+    }
+
+    public void terminate(String executionUnitId) {
+        try {
+            ExecutionUnit<T, R> executionUnit = id2ExecutionUnit.get(executionUnitId);
+            executionUnit.terminate(this);
+        } catch (Exception ex) {
+            log.warn("ExecutionUnit terminate failed, executionId={}", executionUnitId, ex);
+        }
+
+    }
+
+    public void terminateAll() {
+        for (ExecutionUnit<T, R> unit : id2ExecutionUnit.values()) {
+            try {
+                unit.terminate(this);
+            } catch (Exception ex) {
+                log.warn("ExecutionUnit terminate failed, executionId={}", unit.getId(), ex);
+            }
+        }
+    }
+
+    public void skip(String executionUnitId) {
+        ExecutionUnit<T, R> executionUnit = id2ExecutionUnit.get(executionUnitId);
+        executionUnit.skip(this);
+    }
+
+    private void waitForCompletion(@NonNull Set<String> executionIds) throws InterruptedException {
+        while (!Thread.currentThread().isInterrupted()) {
+            List<ExecutionResult<R>> incompleteResults = executionIds.stream()
+                    .map(this::getExecutionResult).filter(result -> !result.isCompleted())
+                    .collect(Collectors.toList());
+            if (CollectionUtils.isEmpty(incompleteResults)) {
+                break;
+            }
+            try {
+                Thread.sleep(500L);
+            } catch (InterruptedException e) {
```

Comment:
use flag to represent isInterrupted call,  isInterrupted() call will clear interrupted flag which depends on JVM implementation .  So keep exit flag is more flexible and consistent. It will allow us do resource clean up more elegant。

Reply:
I assume `Thread.currentThread().isInterrupted()` will not reset the flag, but sleep() will throw the InterruptedException and reset the flag to false. In this case, I reset the flag to true explicitly, which is a common approach to make the upper level be aware of the intteruption.